pkg/engine/task/task_apply.go

```diff
@@ -0,0 +1,137 @@
+		return false, nil
+	}
+	return true, nil
+}
+
+// apply method takes a slice of k8s object and applies them using passed client. If an object
+// doesn't exist it will be created. An already existing object will be patched.
+func apply(ro []runtime.Object, c client.Client) ([]runtime.Object, error) {
+	applied := make([]runtime.Object, len(ro))
+
+	for _, r := range ro {
+		key, _ := client.ObjectKeyFromObject(r)
+		existing := r.DeepCopyObject()
+
+		err := c.Get(context.TODO(), key, existing)
+
+		switch {
+		case apierrors.IsNotFound(err): // create resource if it doesn't exist
+			err = c.Create(context.TODO(), r)
+			if err != nil {
+				return nil, err
+			}
+		case err != nil: // raise any error other than StatusReasonNotFound
+			return nil, err
+		default: // update existing resource
+			err := patch(r, existing, c)
+			if err != nil {
+				return nil, err
+			}
+		}
+		applied = append(applied, existing)
+	}
+
+	return applied, nil
+}
+
+// patch calls update method on kubernetes client to make sure the current resource reflects what is on server
+//
+// an obvious optimization here would be to not patch when objects are the same, however that is not easy
+// kubernetes native objects might be a problem because we cannot just compare the spec as the spec might have extra fields
+// and those extra fields are set by some kubernetes component
+// because of that for now we just try to apply the patch every time
+func patch(newObj runtime.Object, existingObj runtime.Object, c client.Client) error {
+	newObjJSON, _ := apijson.Marshal(newObj)
+	key, _ := client.ObjectKeyFromObject(newObj)
+	_, isUnstructured := newObj.(runtime.Unstructured)
+	_, isCRD := newObj.(*apiextv1beta1.CustomResourceDefinition)
+
+	if isUnstructured || isCRD || isKudoType(newObj) {
+		// strategic merge patch is not supported for these types, falling back to merge patch
+		err := c.Patch(context.TODO(), newObj, client.ConstantPatch(types.MergePatchType, newObjJSON))
+		if err != nil {
+			return fmt.Errorf("failed to apply merge patch to object %s: %w", prettyPrint(key), err)
+		}
+	} else {
+		err := c.Patch(context.TODO(), existingObj, client.ConstantPatch(types.StrategicMergePatchType, newObjJSON))
+		if err != nil {
+			return fmt.Errorf("failed to apply StrategicMergePatch to object %s: %w", prettyPrint(key), err)
+		}
+	}
+	return nil
+}
+
+func isKudoType(object runtime.Object) bool {
+	_, isOperator := object.(*v1alpha1.OperatorVersion)
+	_, isOperatorVersion := object.(*v1alpha1.Operator)
+	_, isInstance := object.(*v1alpha1.Instance)
+	return isOperator || isOperatorVersion || isInstance
+}
+
+func isHealthy(ro []runtime.Object, c client.Client) error {
+	for _, r := range ro {
+		err := health.IsHealthy(c, r)
+		if err != nil {
+			key, _ := client.ObjectKeyFromObject(r)
+			return fmt.Errorf("object %s is NOT healthy: %w", prettyPrint(key), err)
+		}
+	}
+	return nil
+}
+
+func prettyPrint(key client.ObjectKey) string {
```

Comment:
I'm not a fan of this being here... the naming is off.   `task.prettyPrint` seems like it would pretty print a task... but this pretty prints a objectkey.   It is private and isn't a big deal, but it seems like it would have utility in other areas of code.

Reply:
I am actually fine with this tiny code duplication, but no strong feelings here :)

Reply:
A common purpose `prettyPrint` method would need a new file, more error handling etc. I decided to keep it small and narrow-purposed for now.